virtcontainers/qemu_ppc64le.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2018 IBM
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+
+	govmmQemu "github.com/intel/govmm/qemu"
+)
+
+type qemuPPC64le struct {
+	// inherit from qemuArchBase, overwrite methods if needed
+	qemuArchBase
+}
+
+const defaultQemuPath = "/usr/libexec/qemu-kvm"
+
+const defaultQemuMachineType = QemuPPC64le
+
+const defaultQemuMachineOptions = "accel=kvm,usb=off"
+
+const defaultPCBridgeBus = "pci.0"
+
+var qemuPaths = map[string]string{
+	QemuPPC64le: defaultQemuPath,
+}
+
+var kernelParams = []Param{
+	{"tsc", "reliable"},
+	{"no_timer_check", ""},
+	{"rcupdate.rcu_expedited", "1"},
+	{"i8042.direct", "1"},
+	{"i8042.dumbkbd", "1"},
+	{"i8042.nopnp", "1"},
+	{"i8042.noaux", "1"},
+	{"noreplace-smp", ""},
+	{"reboot", "k"},
+	{"console", "hvc0"},
+	{"console", "hvc1"},
+	{"iommu", "off"},
+	{"cryptomgr.notests", ""},
+	{"net.ifnames", "0"},
+	{"pci", "lastbus=0"},
+}
+
+var supportedQemuMachines = []govmmQemu.Machine{
+	{
+		Type:    QemuPPC64le,
+		Options: defaultQemuMachineOptions,
+	},
+}
+
+// returns the maximum number of vCPUs supported
+func maxQemuVCPUs() uint32 {
+	return uint32(128)
+}
+
+func newQemuArch(config HypervisorConfig) qemuArch {
+	machineType := config.HypervisorMachineType
+	if machineType == "" {
+		machineType = defaultQemuMachineType
+	}
+
+	return &qemuPPC64le{
+		qemuArchBase{
+			machineType:           machineType,
+			qemuPaths:             qemuPaths,
+			supportedQemuMachines: supportedQemuMachines,
+			kernelParamsNonDebug:  kernelParamsNonDebug,
+			kernelParamsDebug:     kernelParamsDebug,
+			kernelParams:          kernelParams,
+		},
+	}
+}
+
+func (q *qemuPPC64le) capabilities() capabilities {
+	var caps capabilities
+
+	// pseries machine type supports hotplugging drives
+	if q.machineType == QemuPPC64le {
+		caps.setBlockDeviceHotplugSupport()
+	}
+
+	return caps
+}
+
+func (q *qemuPPC64le) bridges(number uint32) []Bridge {
+	var bridges []Bridge
+	var bt bridgeType
+
+	switch q.machineType {
+
+	case QemuPPC64le:
+		bt = pciBridge
+	default:
+		return nil
+	}
+
+	for i := uint32(0); i < number; i++ {
+		bridges = append(bridges, Bridge{
+			Type:    bt,
+			ID:      fmt.Sprintf("%s-bridge-%d", bt, i),
+			Address: make(map[uint32]string),
+		})
+	}
+
+	return bridges
+}
+
+func (q *qemuPPC64le) cpuModel() string {
+	cpuModel := defaultCPUModel
+	if q.nestedRun {
+		cpuModel += ",pmu=off"
+	}
+	return cpuModel
+}
+
+func (q *qemuPPC64le) memoryTopology(memoryMb, hostMemoryMb uint64) govmmQemu.Memory {
+	// NVDIMM device needs memory space 1024MB
+	memoryOffset := 1024
+
+        //  Align hostMemoryMb to 256 MB multiples
```

Comment:
uhhmm this is weird, why 256?, have you tried with 128 MB?

Reply:
Default LMB (Logical Memory Blocksize) in ppc64le arch is 256MB. If we do not adhere to that we get the error: "docker: Error response from daemon: OCI runtime create failed: qemu-system-ppc64: Can't support memory configuration where RAM size 0x80000000 or maxmem size 0x3017400000 isn't aligned to 256 MB: unknown."

Reply:
make sense, thanks @nitkon